File: spring-data-requery-java/src/main/java/com/coupang/springframework/data/requery/repository/config/EnableRequeryRepositories.java
package com.coupang.springframework.data.requery.repository.config;

import com.coupang.springframework.data.requery.repository.support.RequeryRepositoryFactoryBean;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Import;
import org.springframework.data.repository.config.DefaultRepositoryBaseClass;
import org.springframework.data.repository.query.QueryLookupStrategy;

import java.lang.annotation.*;

/**
 * EnableRequeryRepositories
 *
 * @author debop@coupang.com
 * @since 18. 6. 4
 */
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@Documented
@Inherited
@Import(RequeryRepositoriesRegistar.class)
public @interface EnableRequeryRepositories {

    String[] value() default {};

    String[] basePackages() default {};

    Class<?>[] basePackageClasses() default {};

    ComponentScan.Filter[] includeFilters() default {};

    ComponentScan.Filter[] excludeFilters() default {};

    String repositoryImplementationPostfix() default "Impl";

    String namedQueriesLocation() default "";

    QueryLookupStrategy.Key queryLookupStrategy() default QueryLookupStrategy.Key.CREATE_IF_NOT_FOUND;

    Class<?> repositoryFactoryBeanClass() default RequeryRepositoryFactoryBean.class;

    Class<?> repositoryBaseClass() default DefaultRepositoryBaseClass.class;

    String transactionManagerRef() default "transactionManager";

    boolean considerNestedRepositories() default false;

    boolean enableDefaultTransactions() default true;
}

File: spring-data-requery/src/main/kotlin/org/springframework/data/requery/repository/config/EnableRequeryRepositories.kt
package org.springframework.data.requery.repository.config

import org.springframework.context.annotation.ComponentScan
import org.springframework.context.annotation.Import
import org.springframework.data.repository.config.DefaultRepositoryBaseClass
import org.springframework.data.repository.query.QueryLookupStrategy
import org.springframework.data.requery.repository.support.RequeryRepositoryFactoryBean
import java.lang.annotation.Inherited
import kotlin.reflect.KClass

/**
 * org.springframework.data.requery.repository.config.EnableRequeryRepositories
 *
 * @author debop
 */

@Target(AnnotationTarget.CLASS)
@Retention(AnnotationRetention.RUNTIME)
@MustBeDocumented
@Inherited
@Import(RequeryRepositoriesRegistar::class)
annotation class EnableRequeryRepositories(

    vararg val value: String = [],

    val basePackages: Array<String> = [],

    val basePackageClasses: Array<KClass<*>> = [],

    val includeFilters: Array<ComponentScan.Filter> = [],

    val excludeFilters: Array<ComponentScan.Filter> = [],

    val repositoryImplementationPostfix: String = "Impl",

    // Not supported 
    //    val namedQueriesLocation: String = "",

    val queryLookupStrategy: QueryLookupStrategy.Key = QueryLookupStrategy.Key.CREATE_IF_NOT_FOUND,

    val repositoryFactoryBeanClass: KClass<*> = RequeryRepositoryFactoryBean::class,

    val repositoryBaseClass: KClass<*> = DefaultRepositoryBaseClass::class,

    val transactionManagerRef: String = "transactionManager",

    val considerNestedRepositories: Boolean = false,

    val enableDefaultTransactions: Boolean = true
)
